src/ZstdSharp/SafeHandles.cs

```diff
@@ -0,0 +1,158 @@
+﻿using System;
+using System.Runtime.InteropServices;
+using ZstdSharp.Unsafe;
+
+namespace ZstdSharp
+{
+    /// <summary>
+    /// Provides the base class for ZstdSharp <see cref="SafeHandle"/> implementations.
+    /// </summary>
+    /// <remarks>
+    /// Even though ZstdSharp is a managed library, its internals are using unmanaged
+    /// memory and we are using safe handles in the library's high-level API to ensure
+    /// proper disposal of unmanaged resources and increase safety.
+    /// </remarks>
+    /// <seealso cref="SafeCctxHandle"/>
+    /// <seealso cref="SafeDctxHandle"/>
+    internal abstract unsafe class SafeZstdHandle : SafeHandle
+    {
+        /// <summary>
+        /// Parameterless constructor is hidden. Use the static <c>Create</c> factory
+        /// method to create a new safe handle instance.
+        /// </summary>
+        protected SafeZstdHandle() : base(IntPtr.Zero, true)
+        {
+        }
+
+        public sealed override bool IsInvalid => handle == IntPtr.Zero;
+    }
+
+    /// <summary>
+    /// Safely wraps an unmanaged Zstd compression context.
+    /// </summary>
+    internal sealed unsafe class SafeCctxHandle : SafeZstdHandle
+    {
+        /// <inheritdoc/>
+        private SafeCctxHandle()
+        {
+        }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="SafeCctxHandle"/>.
+        /// </summary>
+        /// <returns></returns>
+        /// <exception cref="ZstdException">Creation failed.</exception>
+        public static SafeCctxHandle Create()
+        {
+            var safeHandle = new SafeCctxHandle();
+            bool success = false;
+            try
+            {
+                var cctx = Methods.ZSTD_createCCtx();
+                if (cctx == null)
+                    throw new ZstdException(ZSTD_ErrorCode.ZSTD_error_GENERIC, "Failed to create cctx");
+                safeHandle.SetHandle((IntPtr)cctx);
```

Comment:
Can we create a safe handle here? When we already have a valid `cctx`.

Reply:
I am following the pattern used by the P/Invoke source generator. Here's the code it generates in .NET 7 (it generates equivalent code in .NET 8 but less human-readable):

```csharp
// User-provided code:
internal partial class Program
{
    [LibraryImport("foo")]
    public static partial SafeFileHandle Foo();
}

// Generated code (with full type qualifications shortened)
// <auto-generated/>
internal unsafe partial class Program
{
    [GeneratedCode("Microsoft.Interop.LibraryImportGenerator", "7.0.10.16914")]
    [SkipLocalsInit]
    public static partial SafeFileHandle Foo()
    {
        bool __invokeSucceeded = default;
        SafeFileHandle __retVal;
        IntPtr __retVal_native = default;
        // Setup - Perform required setup.
        __retVal = new SafeFileHandle();
        try
        {
            {
                __retVal_native = __PInvoke();
            }

            __invokeSucceeded = true;
        }
        finally
        {
            if (__invokeSucceeded)
            {
                // GuaranteedUnmarshal - Convert native data to managed data even in the case of an exception during the non-cleanup phases.
                Marshal.InitHandle(__retVal, __retVal_native);
            }
        }

        return __retVal;
        // Local P/Invoke
        [DllImport("foo", EntryPoint = "Foo", ExactSpelling = true)]
        static extern unsafe IntPtr __PInvoke();
    }
}
```

We could make it simpler, but following .NET's pattern gives some assurance that we are not inadvertently causing a leak.